course_UvA-DL/inception-resnet-densenet/Inception_ResNet_DenseNet.py

```diff
@@ -0,0 +1,1298 @@
+# Throughout this tutorial, we will train and evaluate the models on the CIFAR10 dataset.
+# This allows you to compare the results obtained here with the model you have implemented in the first assignment.
+# As we have learned from the previous tutorial about initialization, it is important to have the data preprocessed with a zero mean.
+# Therefore, as a first step, we will calculate the mean and standard deviation of the CIFAR dataset:
+
+# %%
+train_dataset = CIFAR10(root=DATASET_PATH, train=True, download=True)
+DATA_MEANS = (train_dataset.data / 255.0).mean(axis=(0, 1, 2))
+DATA_STD = (train_dataset.data / 255.0).std(axis=(0, 1, 2))
+print("Data mean", DATA_MEANS)
+print("Data std", DATA_STD)
+
+# %% [markdown]
+# We will use this information to define a `transforms.Normalize` module which will normalize our data accordingly.
+# Additionally, we will use data augmentation during training.
+# This reduces the risk of overfitting and helps CNNs to generalize better.
+# Specifically, we will apply two random augmentations.
+#
+# First, we will flip each image horizontally by a chance of 50% (`transforms.RandomHorizontalFlip`).
+# The object class usually does not change when flipping an image, and we don't expect any image information to be dependent on the horizontal orientation.
+# This would be however different if we would try to detect digits or letters in an image, as those have a certain orientation.
+#
+# The second augmentation we use is called `transforms.RandomResizedCrop`.
+# This transformation scales the image in a small range, while eventually changing the aspect ratio, and crops it afterward in the previous size.
+# Therefore, the actual pixel values change while the content or overall semantics of the image stays the same.
+#
+# We will randomly split the training dataset into a training and a validation set.
+# The validation set will be used for determining early stopping.
+# After finishing the training, we test the models on the CIFAR test set.
+
+# %%
+test_transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize(DATA_MEANS, DATA_STD)])
+# For training, we add some augmentation. Networks are too powerful and would overfit.
+train_transform = transforms.Compose([
+    transforms.RandomHorizontalFlip(),
+    transforms.RandomResizedCrop((32, 32), scale=(0.8, 1.0), ratio=(0.9, 1.1)),
+    transforms.ToTensor(),
+    transforms.Normalize(DATA_MEANS, DATA_STD)
+])
+# Loading the training dataset. We need to split it into a training and validation part
+# We need to do a little trick because the validation set should not use the augmentation.
+train_dataset = CIFAR10(root=DATASET_PATH, train=True, transform=train_transform, download=True)
+val_dataset = CIFAR10(root=DATASET_PATH, train=True, transform=test_transform, download=True)
+set_seed(42)
+train_set, _ = torch.utils.data.random_split(train_dataset, [45000, 5000])
+set_seed(42)
+_, val_set = torch.utils.data.random_split(val_dataset, [45000, 5000])
```

Comment:
```suggestion
set_seed(42)
train_set, val_set = torch.utils.data.random_split(train_dataset, [45000, 5000])
```

Reply:
The `val_set` uses the `val_dataset` to split the dataset from, and the `train_set` splits the `train_dataset`. The two different datasets apply different transformations, which is why I couldn't merge them. Setting the transformations afterwards lead to issues in the data loader, hence this sub-optimal solution.